File: tools/fase_a_refactor.mjs
import fs from 'node:fs';
import path from 'node:path';
import {parse} from '@babel/parser';
import traverseModule from '@babel/traverse';
import generateModule from '@babel/generator';

const traverse=traverseModule.default||traverseModule;
const generate=generateModule.default||generateModule;
const ROOT=process.cwd();
const TEXTOS=path.join(ROOT,'datos','textos.json');
const FILES=['app.js','averia.js','fase5.js','fase6.js','perfiles.js'].map(n=>path.join(ROOT,'build',n));
const TPL=path.join(ROOT,'build','plantilla.html');
const ACC=/[áéíóúñüÁÉÍÓÚÑÜ¿¡]/;
const SPANISH=/\b(?:semana|sesion|servidor|progreso|copia|archivo|guardar|abrir|cerrar|borrar|eliminar|respuesta|pregunta|tarjeta|nueva|hoy|dias|dia|recuperacion|robustez|herramientas|datos|estado|salida|revisar|copiar|perfil|cuentas|curso|avance|marcado|marcar|dominado|dominadas|repaso|fallo|fallar|validado|importado|exportado|bloque|actual|total|desde|hasta|todavia|puedes|puede|quieres|quiero|ninguna|ningun|ninguno|primero|ultima|ultimo|aqui|atras|delante|vacio|disponible)\b/i;
const data=JSON.parse(fs.readFileSync(TEXTOS,'utf8'));

function flatten(x,p='',out={}){
  if(x&&typeof x==='object'&&!Array.isArray(x)){
    for(const [k,v] of Object.entries(x)){
      if(k.startsWith('_'))continue;
      const q=p?`${p}.${k}`:k;
      flatten(v,q,out);
    }
  }else if(typeof x==='string')out[p]=x;
  return out;
}
let flat=flatten(data),rev=new Map(Object.entries(flat).map(([k,v])=>[v,k]));
const ren=data._renombres||{};
const counters={};
const marker=k=>`__T:${k}__`;

function applyRenames(v){
  let s=v;
  const pairs=Object.entries(ren).filter(([k])=>!k.startsWith('_')).sort((a,b)=>b[0].length-a[0].length);
  for(const [old,nuevo] of pairs)if(s===old)return nuevo;
  s=s.replace(/Tu regreso/g,ren['Tu regreso']||'Mi progreso');
  s=s.replace(/Regreso/g,ren.Regreso||'Mi progreso');
  s=s.replace(/Fase final/g,ren['Fase final']||'Cierre del curso');
  s=s.replace(/Bitácora y recuperación/g,ren['Bitácora y recuperación']||'Bitácora');
  s=s.replace(/Autocorrector/g,ren.Autocorrector||'Revisar salida');
  s=s.replace(/Modo Avería/g,ren.Avería||'Practicar averías');
  return s;
}
function looksCode(q){
  if(/https?:\/\//.test(q))return true;
  if(/^([.#/\[{]|--)/.test(q)&&!(/\s/.test(q)))return true;
  if(/^[a-z][A-Za-z0-9_.:/#@-]*$/.test(q))return true;
  if((q.includes(';')&&q.includes(':'))||/^[A-Za-z-]+\s*:/.test(q))return true;
  if(/^cursoVps[A-Za-z0-9]+$/.test(q))return true;
  return false;
}
function candidate(v){
  const q=String(v).trim();
  if(!q||q.includes('__T:')||!/[A-Za-zÁÉÍÓÚÑÜáéíóúñü¿¡]/.test(q))return false;
  if(looksCode(q))return false;
  if(['true','false','null','undefined','object','boolean','string','number','open','active','main','text','bash','html','ini','nginx','yaml','dockerfile','es-ES','GET','POST'].includes(q))return false;
  if(ACC.test(q))return true;
  if(/\s/.test(q))return true;
  return /^[A-Z][A-Za-z]+$/.test(q);
}
function ensure(value,filekey){
  const cleaned=applyRenames(String(value).trim());
  if(rev.has(cleaned))return rev.get(cleaned);
  counters[filekey]=(counters[filekey]||0)+1;
  const n=String(counters[filekey]).padStart(3,'0');
  data.extra=data.extra||{};
  data.extra[filekey]=data.extra[filekey]||{};
  while(Object.prototype.hasOwnProperty.call(data.extra[filekey],n)){
    counters[filekey]++;
    return ensure(value,filekey);
  }
  data.extra[filekey][n]=cleaned;
  const k=`extra.${filekey}.${n}`;
  flat[k]=cleaned;rev.set(cleaned,k);
  return k;
}
function splitSentinels(s){return s.split(/(\u0001E\d+\u0002)/g).filter(x=>x!=='')}
function textPiece(s,filekey,aggressive=false){
  const lead=s.match(/^\s*/)?.[0]||'',trail=s.match(/\s*$/)?.[0]||'',core=s.slice(lead.length,s.length-trail.length);
  const human=aggressive?Boolean(core&&!core.includes('__T:')&&/[A-Za-zÁÉÍÓÚÑÜáéíóúñü¿¡]/.test(core)&&!looksCode(core)):candidate(core);
  if(!human)return s;
  return lead+marker(ensure(core,filekey))+trail;
}
function transformTextWithSentinels(s,filekey,aggressive=false){
  return splitSentinels(s).map(x=>/^\u0001E\d+\u0002$/.test(x)?x:textPiece(x,filekey,aggressive)).join('');
}
function transformTag(tag,filekey){
  return tag.replace(/\b(aria-label|title|placeholder|alt)=(['"])([\s\S]*?)\2/g,(m,name,q,val)=>`${name}=${q}${transformTextWithSentinels(val,filekey,true)}${q}`);
}
function transformHtmlish(value,filekey){
  let v=applyRenames(String(value));
  if(!(v.includes('<')&&v.includes('>')))return transformTextWithSentinels(v,filekey,true);
  let out='',i=0;
  while(i<v.length){
    if(v[i]==='<'){
      const j=v.indexOf('>',i+1);
      if(j<0){out+=v.slice(i);break}
      out+=transformTag(v.slice(i,j+1),filekey);i=j+1;
    }else{
      const j=v.indexOf('<',i),end=j<0?v.length:j;
      out+=transformTextWithSentinels(v.slice(i,end),filekey,true);i=end;
    }
  }
  return out;
}
function skipCss(pathObj){
  const p=pathObj.parentPath;
  return Boolean(p?.isVariableDeclarator?.()&&p.node.id?.type==='Identifier'&&/css|style/i.test(p.node.id.name));
}
function refactorJs(file){
  const filekey=path.basename(file,'.js'),src=fs.readFileSync(file,'utf8');
  const ast=parse(src,{sourceType:'script',allowReturnOutsideFunction:true,plugins:['optionalChaining','nullishCoalescingOperator']});
  traverse(ast,{
    StringLiteral(p){
      if(p.parentPath?.isObjectProperty?.()&&p.parentPath.node.key===p.node&&!p.parentPath.node.computed)return;
      const value=p.node.value;
      if(value.includes('<')&&value.includes('>')){
        const nv=transformHtmlish(value,filekey);if(nv!==value)p.node.value=nv;
      }else if(candidate(value))p.node.value=marker(ensure(value,filekey));
    },
    TemplateLiteral(p){
      if(skipCss(p))return;
      const token=i=>`\u0001E${i}\u0002`;let joined='';
      p.node.quasis.forEach((q,i)=>{joined+=(q.value.cooked??q.value.raw);if(i<p.node.expressions.length)joined+=token(i)});
      const nv=transformHtmlish(joined,filekey),parts=nv.split(/\u0001E\d+\u0002/g);
      if(parts.length!==p.node.quasis.length)throw new Error(`No pude separar template en ${file}`);
      p.node.quasis.forEach((q,i)=>{q.value.cooked=parts[i];q.value.raw=parts[i].replace(/`/g,'\\`').replace(/\$\{/g,'\\${')});
    }
  });
  fs.writeFileSync(file,generate(ast,{comments:false,compact:false,retainLines:true,jsescOption:{minimal:true}},src).code+'\n','utf8');
}
function transformHtmlSource(src){
  const filekey='plantilla';let s=src;
  for(const [old,nuevo] of Object.entries(ren).filter(([k])=>!k.startsWith('_')).sort((a,b)=>b[0].length-a[0].length))if(s.includes(old))s=s.split(old).join(marker(ensure(nuevo,filekey)));
  for(const [k,v] of Object.entries(flat).sort((a,b)=>b[1].length-a[1].length))if(v.length>1&&s.includes(v))s=s.split(v).join(marker(k));
  const chunks=s.split(/(<style[\s\S]*?<\/style>|<script[\s\S]*?<\/script>)/ig);
  for(let i=0;i<chunks.length;i++){
    if(/^<(style|script)/i.test(chunks[i]))continue;
    chunks[i]=chunks[i].replace(/\b(aria-label|title|placeholder|alt)=(['"])([\s\S]*?)\2/g,(m,name,q,val)=>`${name}=${q}${transformTextWithSentinels(val,filekey,true)}${q}`);
    chunks[i]=chunks[i].replace(/>([^<>]+)</g,(m,val)=>`>${textPiece(val,filekey,true)}<`);
  }
  s=chunks.join('');
  if(!s.includes('__TEXTOS_JSON__')){
    const re=/<script[^>]*\bid=['"]cursoData['"][^>]*>/i,m=s.match(re);
    if(!m)throw new Error('No encuentro cursoData en plantilla.html');
    const at=m.index;s=s.slice(0,at)+'<script type="application/json" id="textosData">__TEXTOS_JSON__</script>\n'+s.slice(at);
  }
  return s;
}
function cleanForAudit(v){
  return String(v).replace(/__T:[A-Za-z0-9_.-]+__/g,'').replace(/<[^>]*>/g,' ').replace(/\s+/g,' ').trim();
}
function auditJs(file){
  const src=fs.readFileSync(file,'utf8'),ast=parse(src,{sourceType:'script',plugins:['optionalChaining','nullishCoalescingOperator']});
  const residual=[];
  traverse(ast,{
    StringLiteral(p){
      if(p.parentPath?.isObjectProperty?.()&&p.parentPath.node.key===p.node&&!p.parentPath.node.computed)return;
      const q=cleanForAudit(p.node.value);if(q&&!looksCode(q)&&SPANISH.test(q))residual.push(q);
    },
    TemplateLiteral(p){
      if(skipCss(p))return;
      const q=cleanForAudit(p.node.quasis.map(x=>x.value.cooked??x.value.raw).join(' '));if(q&&!looksCode(q)&&SPANISH.test(q))residual.push(q);
    }
  });
  return residual;
}

for(const f of FILES)refactorJs(f);
fs.writeFileSync(TPL,transformHtmlSource(fs.readFileSync(TPL,'utf8')),'utf8');
fs.writeFileSync(TEXTOS,JSON.stringify(data,null,1)+'\n','utf8');

let accents=0,forbidden=[],residual=[];
for(const f of FILES){
  const s=fs.readFileSync(f,'utf8');accents+=(s.match(/[áéíóúñüÁÉÍÓÚÑÜ¿¡]/g)||[]).length;
  for(const old of Object.keys(ren).filter(k=>!k.startsWith('_')))if(s.includes(old))forbidden.push(`${path.basename(f)}:${old}`);
  for(const q of auditJs(f))residual.push(`${path.basename(f)}:${q}`);
}
console.log(`AUDIT_ACCENTS=${accents}`);
console.log(`AUDIT_FORBIDDEN=${forbidden.length}`);
console.log(`AUDIT_RESIDUAL=${residual.length}`);
if(forbidden.length)console.log(forbidden.join('\n'));
if(residual.length)console.log(residual.slice(0,100).join('\n'));
if(accents||forbidden.length||residual.length)process.exitCode=2;
else console.log('FASE_A_REFACTOR_OK');

File: tools/fase_a_refactor.py
#!/usr/bin/env python3
from __future__ import annotations
import json,re
from pathlib import Path

ROOT=Path(__file__).resolve().parents[1]
TEXTOS=ROOT/'datos'/'textos.json'
JS=[ROOT/'build'/n for n in ('app.js','averia.js','fase5.js','fase6.js','perfiles.js')]
TPL=ROOT/'build'/'plantilla.html'
BUILD=ROOT/'build'/'construir.py'
MARK_RE=re.compile(r'__T:([A-Za-z0-9_.-]+)__')
ACC=re.compile(r'[áéíóúñüÁÉÍÓÚÑÜ¿¡]')


def flatten(x,p=''):
    out={}
    if isinstance(x,dict):
        for k,v in x.items():
            if str(k).startswith('_'): continue
            q=f'{p}.{k}' if p else k
            out.update(flatten(v,q))
    elif isinstance(x,str): out[p]=x
    return out

def marker(k): return f'__T:{k}__'

def plain_candidate(s):
    q=s.strip()
    if not q or '__T:' in q or not re.search(r'[A-Za-zÁÉÍÓÚÑÜáéíóúñü¿¡]',q): return False
    if re.search(r'https?://|^[#.[/]|[{};=]|\\[nrt]|^\w+[-_:]\w+$',q): return False
    if q in {'true','false','null','undefined','object','boolean','string','number','open','active','main','text','bash','html','ini','nginx','yaml','dockerfile','es-ES','GET','POST'}: return False
    if ACC.search(q): return True
    if re.search(r'\s',q): return True
    if q[:1].isupper() and re.fullmatch(r'[A-Za-z]+',q): return True
    return False

def html_visible_transform(s,ensure):
    def attr(m):
        before,val,after=m.group(1),m.group(2),m.group(3)
        if plain_candidate(val): val=marker(ensure(val))
        return before+val+after
    s=re.sub(r'((?:aria-label|title|placeholder|alt)=["\'])(.*?)(["\'])',attr,s,flags=re.S)
    def node(m):
        val=m.group(1)
        lead=val[:len(val)-len(val.lstrip())]; trail=val[len(val.rstrip()):]
        core=val.strip()
        if plain_candidate(core): return '>'+lead+marker(ensure(core))+trail+'<'
        return m.group(0)
    return re.sub(r'>([^<>]+)<',node,s)

def transform_js(src,filekey,data):
    flat=flatten(data); rev={v:k for k,v in flat.items()}
    ren=data.get('_renombres',{})
    counter=0
    def ensure(value):
        nonlocal counter,flat,rev
        value=ren.get(value,value)
        if value in rev:return rev[value]
        counter+=1;k=f'_extra.{filekey}.{counter:03d}'
        data.setdefault('_extra',{}).setdefault(filekey,{})[f'{counter:03d}']=value
        flat[k]=value;rev[value]=k
        return k
    # Renombres obligatorios primero.
    for old,new in sorted(((k,v) for k,v in ren.items() if not k.startswith('_')),key=lambda x:len(x[0]),reverse=True):
        if old in src: src=src.replace(old,marker(ensure(new)))
    # Copia ya definida en textos.json.
    for k,v in sorted(flatten(data).items(),key=lambda kv:len(kv[1]),reverse=True):
        if len(v)>=2 and v in src: src=src.replace(v,marker(k))
    # Strings simples.
    str_re=re.compile(r'(["\'])(?P<body>(?:\\.|(?!\1).)*?)\1',re.S)
    def qs(m):
        quote=m.group(1); body=m.group('body')
        if '__T:' in body:return m.group(0)
        if '<' in body and '>' in body:
            nb=html_visible_transform(body,ensure)
            return quote+nb+quote
        if plain_candidate(body): return quote+marker(ensure(body))+quote
        return m.group(0)
    src=str_re.sub(qs,src)
    # Templates: texto visible HTML y fragmentos estáticos alrededor de ${...}.
    temp_re=re.compile(r'`((?:\\.|[^`])*)`',re.S)
    def qt(m):
        body=m.group(1)
        body=html_visible_transform(body,ensure)
        parts=re.split(r'(\$\{[^{}]*\})',body)
        for i,p in enumerate(parts):
            if p.startswith('${') or '__T:' in p: continue
            lead=p[:len(p)-len(p.lstrip())];trail=p[len(p.rstrip()):];core=p.strip()
            # No convertir trozos que son principalmente marcado HTML.
            if core and '<' not in core and '>' not in core and plain_candidate(core):
                parts[i]=lead+marker(ensure(core))+trail
        return '`'+''.join(parts)+'`'
    src=temp_re.sub(qt,src)
    return src

def transform_html(src,data):
    flat=flatten(data);rev={v:k for k,v in flat.items()};ren=data.get('_renombres',{});counter=0
    def ensure(value):
        nonlocal counter
        value=ren.get(value,value)
        if value in rev:return rev[value]
        counter+=1;k=f'_extra.plantilla.{counter:03d}'
        data.setdefault('_extra',{}).setdefault('plantilla',{})[f'{counter:03d}']=value
        rev[value]=k;return k
    for old,new in sorted(((k,v) for k,v in ren.items() if not k.startswith('_')),key=lambda x:len(x[0]),reverse=True):
        src=src.replace(old,marker(ensure(new)))
    for k,v in sorted(flatten(data).items(),key=lambda kv:len(kv[1]),reverse=True):
        if len(v)>=2:src=src.replace(v,marker(k))
    # No tocar CSS ni scripts; solo texto/atributos del HTML estructural.
    chunks=re.split(r'(<style[\s\S]*?</style>|<script[\s\S]*?</script>)',src,flags=re.I)
    for i,ch in enumerate(chunks):
        if re.match(r'<(?:style|script)',ch,re.I): continue
        chunks[i]=html_visible_transform(ch,ensure)
    return ''.join(chunks)

def write_build():
    BUILD.write_text('''#!/usr/bin/env python3\n"""Genera public/index.html autocontenido desde plantilla, aplicaciones y datos validados."""\nfrom __future__ import annotations\nimport html,json,re\nfrom pathlib import Path\nROOT=Path(__file__).resolve().parent.parent\nPLANTILLA=ROOT/'build'/'plantilla.html'; APP=ROOT/'build'/'app.js'; AVERIA=ROOT/'build'/'averia.js'; FASE5=ROOT/'build'/'fase5.js'; FASE6=ROOT/'build'/'fase6.js'; PERFILES=ROOT/'build'/'perfiles.js'; SALIDA=ROOT/'public'/'index.html'\nCURSO=ROOT/'datos'/'curso.json'; PARTES=ROOT/'datos'/'curso_partes'; TEXTOS=ROOT/'datos'/'textos.json'\ndef compactar(x): return json.dumps(x,ensure_ascii=False,separators=(',',':')).replace('</','<\\\\/')\ndef archivo(p): return compactar(json.loads(p.read_text(encoding='utf-8')))\ndef curso():\n    if CURSO.exists(): data=json.loads(CURSO.read_text(encoding='utf-8'))\n    else:\n        ps=sorted(PARTES.glob('curso.*.part'))\n        if len(ps)!=6: raise SystemExit(f'Curso incompleto: se esperaban 6 partes y hay {len(ps)}')\n        try:data=json.loads(''.join(p.read_text(encoding='utf-8') for p in ps))\n        except json.JSONDecodeError as e:raise SystemExit(f'No se pudo reensamblar el curso: {e}') from e\n    assert len(data.get('sesiones',[]))==20, 'Se esperaban 20 sesiones'\n    assert sum(len(s.get('checkpoint',[])) for s in data['sesiones'])==101, 'Se esperaban 101 checkpoints'\n    assert sum(len(s.get('miniQuiz',[])) for s in data['sesiones'])==71, 'Se esperaban 71 preguntas'\n    return compactar(data)\ndef plano(x,p=''):\n    r={}\n    if isinstance(x,dict):\n        for k,v in x.items():\n            if str(k).startswith('_'):continue\n            q=f'{p}.{k}' if p else k;r.update(plano(v,q))\n    elif isinstance(x,str):r[p]=x\n    return r\ndef cargar_textos():return json.loads(TEXTOS.read_text(encoding='utf-8'))\ndef js_escape(s):\n    return ''.join(('\\\\n' if c=='\\n' else '\\\\r' if c=='\\r' else '\\\\t' if c=='\\t' else f'\\\\u{ord(c):04x}' if ord(c)>126 or c in '\\\\"\\\'`$' else c) for c in s)\ndef resolver(src,textos,modo):\n    flat=plano(textos)\n    def sub(m):\n        k=m.group(1)\n        if k not in flat:raise SystemExit(f'Falta texto {k}')\n        return html.escape(flat[k],quote=True) if modo=='html' else js_escape(flat[k])\n    out=re.sub(r'__T:([A-Za-z0-9_.-]+)__',sub,src)\n    if '__T:' in out:raise SystemExit('Quedaron referencias de texto sin resolver')\n    return out\ndef js(p,textos):return resolver(p.read_text(encoding='utf-8'),textos,'js').replace('</script>','<\\\\/script>')\ndef main():\n    textos=cargar_textos();html_src=PLANTILLA.read_text(encoding='utf-8')\n    html=resolver(html_src,textos,'html')\n    rep={'__CURSO_JSON__':curso(),'__REGLAS_JSON__':archivo(ROOT/'datos'/'reglas.json'),'__INSTALA_JSON__':archivo(ROOT/'datos'/'instala.json'),'__RIEL_JSON__':archivo(ROOT/'datos'/'riel.json'),'__TEXTOS_JSON__':compactar(textos),'__APP_JS__':js(APP,textos)}\n    for k,v in rep.items():\n        if k not in html:raise SystemExit(f'Falta marcador {k}')\n        html=html.replace(k,v)\n    extra=''.join('<script>'+js(p,textos)+'</script>' for p in (AVERIA,FASE5,FASE6,PERFILES))\n    if '</body>' not in html:raise SystemExit('La plantilla no contiene </body>')\n    html=html.replace('</body>',extra+'</body>',1)\n    SALIDA.parent.mkdir(parents=True,exist_ok=True);SALIDA.write_text(html,encoding='utf-8');print(f'OK {SALIDA} ({SALIDA.stat().st_size/1024:.1f} KB)')\nif __name__=='__main__':main()\n''',encoding='utf-8')

def ensure_textos_marker(src):
    if '__TEXTOS_JSON__' in src:return src
    tag='<script id="textosData" type="application/json">__TEXTOS_JSON__</script>'
    pos=src.find('<script id="cursoData"')
    if pos<0: raise SystemExit('No encuentro cursoData en plantilla.html')
    return src[:pos]+tag+'\n'+src[pos:]

def audit(data):
    forbidden=[k for k in data.get('_renombres',{}) if not k.startswith('_')]
    bad=[]; accents=0
    for p in JS:
        s=p.read_text(encoding='utf-8')
        accents+=len(ACC.findall(s))
        for old in forbidden:
            if old in s:bad.append(f'{p.name}: renombre prohibido: {old}')
        # Candidatos residuales en strings: reporte conservador.
        for m in re.finditer(r'(["\'])(?P<b>(?:\\.|(?!\1).)*?)\1',s,re.S):
            if plain_candidate(m.group('b')) and '__T:' not in m.group('b') and '<' not in m.group('b'):
                bad.append(f'{p.name}: copia residual: {m.group("b")[:80]}')
    print(f'AUDIT_ACCENTS={accents}')
    print(f'AUDIT_RESIDUAL={len(bad)}')
    for x in bad[:80]:print('  '+x)
    if accents or bad:raise SystemExit('Auditoría de copia no pasó')


def main():
    data=json.loads(TEXTOS.read_text(encoding='utf-8'))
    for p in JS:
        p.write_text(transform_js(p.read_text(encoding='utf-8'),p.stem,data),encoding='utf-8')
    tpl=ensure_textos_marker(TPL.read_text(encoding='utf-8'))
    TPL.write_text(transform_html(tpl,data),encoding='utf-8')
    TEXTOS.write_text(json.dumps(data,ensure_ascii=False,indent=1)+'\n',encoding='utf-8')
    write_build()
    audit(data)
    print('FASE_A_OK')

if __name__=='__main__':main()
